pyquil/api/_qpu.py

```diff
@@ -122,24 +122,23 @@ def _build_client(self) -> Client:
 
     @property
     def client(self) -> Client:
-        if self.session:
-            if not (self.session.config.get_engagement()
-                    and self.session.config.get_engagement().is_valid()
-                    and self._client):
-                self._client = self._build_client()
-        elif not self._client:
+        """
+        Return a memoized client with a valid engagement if it exists; otherwise, build a new one
+        and return it. If this QPU object does not have a ForestSession, then engagement is not
+        applicable.
+        """
+        if not (self._client and (self.session is None or self.session.config.engagement)):
```

Comment:
where did the `is_valid` part go?

Reply:
no invalid engagement will be returned by `config.engagement`, validity is checked within `session`. There's a change to this en route though